### PR TITLE
Add ToolTip class

### PR DESCRIPTION
**Pull request by Moosems**

### CODE REVIEW

**Comment by Moosems** [comment]
@rdbende Can you help sort out the imports issue?

**Comment by rdbende** [comment]
I'll rewiew it asap.

**Comment by rdbende** [comment]
Just looked at the CI results, neither the black or the isort error is from you code 👍

**Comment by Moosems** [comment]
So what do I/we do?

**Comment by rdbende** [comment]
I will review it as I can. I also have some ideas for it. Maybe the `bg` and `fg` colors aren't necessary, and the tooltip should use a style instead, so it can be themed.

**Comment by rdbende** [comment]
I have an other idea. Since many `ToolTip` objects can consume a lot of memory, I thought it would be better to have a single tooltip provider object that displays the tooltips where needed. This also avoids the bug where two tooltips appear at the same time for some reason.

File: tukaan/widgets/tooltip.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 from tukaan import Label
+from tukaan._tcl import Tcl
```
Comment (rdbende) [inline tukaan/widgets/tooltip.py:26]: Just noticed, `isort` had an issue with this line. It likes two blank lines between imports and the start of the actual code, that was its problem.

File: tukaan/widgets/tooltip.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+class ToolTip():
+    def __init__(self, window, widget=None, orient="E", *args, **kwargs):
+        try:
+            self.text = kwargs.pop("text")
+        except:
+            raise Exception("No text provided for ToolTip")
+        self.window = window
+        self.widget = widget
+        self.orient=orient
```
Comment (rdbende) [inline tukaan/widgets/tooltip.py]: Black complained about this line, it likes to have spaces around operators.

If you're working in your local clone, you can `pip install black isort`, and run them on your files `black tukaan/*.py tukaan/*/*.py` (same with `isort`) to format them, this way the checks won't fail.

**Review by rdbende — APPROVED** [review]

**Comment by rdbende** [comment]
Looks good to me!

image

File: tukaan/widgets/tooltip.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        owner_x = Tcl.call(int, "winfo", "rootx", widget)
+        owner_width = Tcl.call(int, "winfo", "reqwidth", widget)
+        tip_width = Tcl.call(int, "winfo", "reqwidth", ".tooltip")
+        tip_x = owner_x + owner_width // 2 - tip_width // 2
+
+        owner_y = Tcl.call(int, "winfo", "rooty", widget)
+        tip_height = Tcl.call(int, "winfo", "reqheight", ".tooltip")
+        tip_y = owner_y - tip_height - 5
+
+        if tip_y <= Tcl.call(int, "winfo", "rooty", ".app"):
```
Comment (rdbende) [inline tukaan/widgets/tooltip.py:96]: Upsie, this shouldn't be `.app`, rather the parent toplevel.

**Comment by Moosems** [comment]
Wow quite a few changes! I love how now it's a part of the parent widget instead of being it's own individual thing! What does it look like to create a tooltip now? Can they still specify which side it should be on?

**Comment by rdbende** [comment]
```python
button = tukaan.Button(parent, text="Click me!", tooltip="Button")  # create button with tooltip

button.tooltip = "New text"  # change tooltip text
```

Currently the side is automatic. By default the tooltip is displayed above the widget, but if the tooltip were to leave the parent window and cover the titlebar, it will automatically appear below the widget instead.
I plan to make some configuration options, but they would be global configs, and there'd be still some automatic side detection, to ensure that the tooltip is always displayed at the right position.

**Comment by Moosems** [comment]
Ok. I personally like the tool tips to the left of the widget so that's why I added the option to move it because it isn't a popular opinion. Could there be an option to specify like `tooltip=("A button", "E")`?

**Review by rdbende — APPROVED** [review]

**Pull request merged**